### PR TITLE
Add triggers for variants

### PR DESCRIPTION
**Pull request by bradphilips**
This request is to add triggers for variants to control the bucket segments based on data gathered from other services/application data instead of weighted variants. This will allow for A/B testing based off of controlled segments.

### CODE REVIEW

**Comment by gingerlime** [comment]
Thanks for your contribution @bradphilips. Sorry, but I'm not sure I fully understand the use-case here. Can you give an example of when this is necessary?

**Comment by bradphilips** [comment]
@gingerlime, Sure thing and no problem it's probably something that doesn't come up much because a lot of A/B test tools are mostly driven by random distribution. The use case here is actually fairly simple. Basically, we want to allow for the person defining the A/B test to control the group of customers that get a specific variant based on attributes of a customer (or a targeted segmentation group) for that specific A/B test. 

**For example (a little contrived Amazon example):**

For all customers that have signed up for one-click pay, Amazon wants to participate in an A/B test for _slide to pay_. This is defined by the overall trigger for the experiment.

We **also** only want customers that use the one-click feature over 10 times a month to be selected for the _A_ variant which is a blue swipe to pay.

All others customers will get a _B_ experience which is an orange swipe to pay.

This is where the triggers for the variants come into play. With this, we would define our AlephBet A/B test as follows:

```javascript
var swipeToPayExperiment = new AlephBet.Experiment({
  'name': 'Swipe One-Click Pay',
  'trigger': function() {
    return customerAttributes.isOneClickCustomer;
  },
  'variants': {
      'A': {
          trigger: function() { return customerAttributes.usedOneClickInLastMonth > 10; },
          activate: function() { $("#oneclick").toggleClass("blueSlider"); }
      },
      'B': {
        trigger: function() { return customerAttributes.usedOneClickInLastMonth <= 10; },
        activate: function() { $("#oneclick").toggleClass("orangeSlider"); }
      }
  },
});
```

Basically, this allows the A/B test buckets (variants) to be determined by a pre-defined set of parameters as opposed to weighted in a random group that could go to either A or B based off the weighted distribution. Also, it allows for mix-adjusting the segments or changing them based on variations in the data's origin -- for example a customer from Ohio vs a customer in California.

Furthermore, this allows for more targeted marketing because with backend data we already know which users will fall into these groups and provide an email or push notification based on their specific experience. 

Let me know if this makes sense. Thanks!

**Comment by gingerlime** [comment]
Thanks for taking the time to explain, @bradphilips. I think I understand now.

I have a bit of a problem with this approach however. First of all, this isn't really A/B testing any more. You're essentially using Alephbet to feature flag certain behaviour + track certain goals. The feature flagging aspects can be easily controlled without Alephbet. i.e. instead of setting up an experiment, you just need a simple condition. e.g. for your example

```javascript
if (customerAttributes.isOneClickCustomer) {
  if (customerAttributes.usedOneClickInLastMonth > 10) {
    $("#oneclick").toggleClass("blueSlider")
  } else {
    $("#oneclick").toggleClass("orangeSlider")
  }
}
```

The goal tracking might be interesting, but can also be achieved without Alephbet... ?

Don't get me wrong, I'm really keen to get people to use Alephbet, and even more so when someone contributes code. Just not entirely sure if this is the best use-case for the library.

That being said, I was thinking of a way of forcing a particular variant. This is mostly meant to be used when manually checking the experiment / testing. That's a somehow more "appropriate" use case in my opinion (and of course, can be (ab)used any way people decide). Doing that however, I'm not sure whether adding a `trigger` method for each variant is the right way. Maybe some kind of an override function? or allowing to pass a callback that would determine which variant is picked? Haven't thought about the API for this use case, but something in this direction.

What do you think?

**Comment by bradphilips** [comment]
@gingerlime , sorry so late on getting back to this. I see some of your points but I disagree that its a simple feature flag -- the goal of this is to have the ability to let another server control the distribution between a set of variants. The goal of the A/B test is still the same, we want to measure which experience got the best results and cache experiences for return customers while randomizing the experience for new users. 

The big difference here is that the distributions can be associated with a particular user and/or based on random distribution from information we gather and store elsewhere; such as an API call to integrate the same experience across mobile and web offerings of the same A/B test for example. 

I definitely think the override may help if you feel that's more appropriate. I think the only concern I might see there is the complexity in picking the variant if its composed all in the same method as opposed to providing a conditional for each variant. 
Thanks!

**Comment by gingerlime** [comment]
Thanks @bradphilips.

> The goal of the A/B test is still the same, we want to measure which experience got the best results and cache experiences for return customers while randomizing the experience for new users.

This you can already do with Alephbet. There's built in support for [user-based cross device tracking](#16)

Only forcing a particular variant isn't supported, and this feels more like something feature flags can do for you if you anyway rely on a server or something else to make this decision for you. I understand that you want to track goals and be able to compare them for each variant. You can certainly use Alephbet for that, but you can also interface directly with your Tracking Backend or any other backend... Yes, it would involve a bit more code to store goals and send them, and I can therefore see the appeal of Alephbet, which already does that.

Adding a trigger to each variant feels a bit counter-intuitive to me, even though I can understand that it might be simpler to use than an override method. I'll have to think about. Maybe others can share their opinions? @joker-777 @szymansd ?

(btw, the `user_id` method can actually allow you to "hack" a variant chooser already. Instead of returning an id, you can find values that end up picking the variant you want, and then use this method to make the variant decision for you... definitely a ugly hack rather than a real solution)

**Comment by joker-777** [comment]
@bradphilips Can you explain, why exactly would you like to make the differentiation between the usage amount of onClick? Usually, you just exclude certain users who already have seen a certain variant and apply the a/b test on "fresh" users.
Also, what is the benefit of showing one variant to only users of one state and another variant to users of a different state? If one variant wins, would you show it then to just the users of this particular state or to all users?

**Comment by bradphilips** [comment]
@joker-777, not fully sure what you mean by the `usage amount of onClick`. In my contrived example, I am merely segmenting the variants based on conditions of the customer segment. This is the same as random distribution (weighting 20% to bucket A, 80% to bucket B) its just identifying these user groups by a pre-defined subset and not as random as the random number distribution Alephbet currently provides. In my example, this is defined as using one click more than 10 times in the prior month. This could also be a segmentation group already known or a custom randomize function that assigns these users to the variant.

We would still follow all the same rules of an A/B test, for example showing it to only fresh users, measuring the goals of these variants and assigning a winner.

**Comment by gingerlime** [comment]
@bradphilips another thing that I just realised... You can already trigger specific variants using @szymansd's `weight` options actually, e.g.

```javascript
var swipeToPayExperiment = new AlephBet.Experiment({
  name: 'Swipe One-Click Pay',
  trigger: function() {
    return customerAttributes.isOneClickCustomer;
  },
  variants: {
      'A': {
        weight: customerAttributes.usedOneClickInLastMonth > 10 ? 1 : 0,
        activate: function() { $("#oneclick").toggleClass("blueSlider"); }
      },
      'B': {
        weight: customerAttributes.usedOneClickInLastMonth > 10 ? 0 : 1,
        activate: function() { $("#oneclick").toggleClass("orangeSlider"); }
      }
  },
});
```

Essentially, all you have to do is set `weight` to `0` for all variants that you don't want to pick, and any positive integer value for the variant you want to pick...

**Comment by bradphilips** [comment]
@gingelime, good call. Didn't realize this was an option. Thank you! ☺️

**Comment by gingerlime** [comment]
Yeah, I also didn't think about it, even though `weight`s were only recently added. In any case, it seems to fit quite well to this use-case as well, which I think is pretty great :)

I'll close this one, but if there's any other idea for improvement, please feel free.
